Added bipolar membrane unit model and costing (#1500)
watertap-org/watertap: docs/technical_reference/unit_models/Biploar_electrodialysis_0D.rst

> Bipolar Electrodialysis (0D)

Reviewer: This is great! While we're here, you should also add an rst file to describe costing, e.g.,: https://watertap.readthedocs.io/en/latest/technical_reference/costing/detailed_unit_model_costing.html

Author: That's a good idea. However, I made some changes to the unit model costing and would like some feedback (before committing it to the docs).
While doing analysis I realised that a big component (negative cost/value addition) of the bipolar costing might be best done on the flowsheet level. So, I have removed it (in the most recent version).

As it stands the BPMED costing is very similar to the ED costing. The difference being the variable name for number of membranes. (cell_pair_num vs cell_num). Is there a way to inherit the ED costing?

Also, just for future reference I plan to add the bpmed + ED unit model (a configuration that is commonly used and something that I have been using for analysis). This would require minor alteration to the costing. It would be nice to know the capabilities/limits of the inheritance. Since that would inform whether I need a costing just for bipolar membrane. 
I am happy to elaborate more on this issue.

> ============================

Reviewer: Looks like you updated the terminology to acidic and basic channels. One other place where this should be updated is in the png diagram.

Author: Thanks for pointing that out. Somehow I missed that.

Introduction
------------

Bipolar electrodialysis, an electrochemical separation technology, has primarily been used to generate acids and bases
from waste salts produced in water purification. By providing in-situ access to valuable raw materials bipolar membranes can
significantly reduce the total cost of the operation.  This cell stack is shown in Figure 1 with **basate** and **acidate** channels, that produce base and acid
respectively. More overview of the bipolar
electrodialysis technology can be found in the *References*.

.. figure:: ../../_static/unit_models/BPEDdiagram.png
    :width: 600
    :align: center

    Figure 1. Schematic representation of a bipolar electrodialysis unit

> One bipolar membrane along with the **Acidate** and **Basate** channels can thus be treated as a modelling unit that can

Reviewer: Are acidate and basate typical terminology for BPMD?

Author: They are not. I needed a name for the acid & base channel. 
I went off diluted/ concentrated channel which are called diluate/concentrate (in ED documentation). Extrapolating from this I chose acidate/basate.

If anybody has better suggestion I am happy to use it.

Reviewer 2: In theory they are an acid concentrate and base concentrate (increasing in concentration), where the brine is a brine diluate (deceasing in concentration) based on the donation of ions toward the objective of the system, correct?

Author: Yes, they do increase concentration. But I think there is a subtle difference. BPMED increase concentration via production of new acid/base (through water splitting) and not by just transporting ions across the membranes. I feel like that deserved a distinction in nomenclature. Thoughts?

Reviewer: I'd go with either acid and base or acidic and basic.

Author: @adam-a-a That's reasonable. I have changed the names to Acidic & Basic

multiply to larger-scale systems. The presented bipolar electrodialysis model establishes mathematical descriptions of
ion and water transport across the membrane along with water splitting. Modelled transfer mechanisms include
electrical migration, diffusion of ions, osmosis, electroosmosis, and water splitting. The following are the key

> assumptions made:

Reviewer: This sounds great. A question that comes to mind is -- what notable mechanisms does the model _not_ account for at this time (but technically could be based on latest literature)? I suggest creating an issue to track any such mechanisms so that we chart a path for future enhancement, whether that be by you or another developer later on.

Author: @adam-a-a This is a good question. I actually faced the other problem. It was challenging finding the appropriate steady state equations for bipolar membrane (there are many equations/features  for dynamic bipolar modelling but that's a can of worms for another day). 
Even now there are some features (like the boundary layer details) that have not been added to the unit model because I can't find appropriate equations.

* The **acidic** and **basic** side channels have identical geometry.
* For each channel, component fluxes are uniform in the bulk solutions (the 0-dimensional assumption)  and are set as the average of inlet and outlet of each channel.

> * Steady state: all variables are independent on time.

Reviewer: ```suggestion
* Steady state: all variables are independent of time.
```

* Co-current flow operation. 
* Ideality assumptions: activity, osmotic, and van't Hoff coefficients are set at one.
* All ion-exchange membrane properties (ion and water transport number, resistance, permeability) are constant.
* Detailed concentration gradient effect at membrane-water interfaces is neglected. 
* Constant pressure and temperature through each channel. 

Control Volumes
---------------

This model has two control volumes for the acidic and basic channels.

* **acidic** channel

> * **basic** side channel

Reviewer: Is one called `acidic` and the other called `basic` or `basic_side`? I'd remove "side" if not used in the attribute name to avoid confusion.

( H\ :sub:`2`\ O, neutral solutes, and ions, including Proton and Hydroxide ion) form a Pyomo set in the model.
For a clear model demonstration, this document uses a NaCl water solution along with the products of water splitting, H\ :sup:`+` and OH\ :sup:`-`, hereafter.

This model can mathematically take a multi-component (i.e., one salt molecule to be treated) as an input; nevertheless
a multi-component solution creates unknown or difficult-to-specify parameters, e.g., the electrical transport numbers through membranes,
the multi-ion diffusivity, etc., and physical relationships, which may result in ill-posed or ill-conditioned problems challenging the models'
numerical solutions.  While we continuously work on advancing our models to absorb new principles revealed by progressing
research, we advise the users be very **cautious** with simulating multi-component systems by this programmed model for aspects stated above.

.. csv-table:: **Table 1.** List of Set
   :header: "Description", "Symbol", "Indices"

   "Time", ":math:`t`", "[t] ([0])\ :sup:`1`"
   "Phase", ":math:`p`", "['Liq']"

>    "Component", ":math:`j`", "['H\ :sub:`2` \O', 'Na\ :sup:`+`', 'Cl\ :sup:`-`', 'H\ :sup:`+`', 'OH\ :sup:`-`']"

Reviewer: It is not yet clear whether this unit works with MCAS, the NaCl prop package, both, and/or other property models. Can you elaborate?

Author: How do I check this? I took what the ED folks have been using and ran with it. So, theoretically whatever is applicable for the prop packages in the ED unit model should be applicable here (with the caveat that users should also supply H_+ and OH_- ).

   "Ion", ":math:`j`", "['Na\ :sup:`+`', 'Cl\ :sup:`-`', 'H\ :sup:`+`', 'OH\ :sup:`-`'] \  :sup:`2`"
   "Membrane", "n/a", "['bpem']"

**Notes**
 :sup:`1` The time set index is set as [0] in this steady-state model and is reserved majorly for the future extension
 to a dynamic model.

 :sup:`2` "Ion" is a subset of "Component" and uses the same symbol j.

Degrees of Freedom
------------------
The bipolar membrane model has multiple degrees of freedom, among which temperature, pressure, and component molar flow
rate are state variables that are fixed as initial conditions. The rest are parameters that should be provided in order
to fully solve the model. The exact degrees of freedom depend on the mode of operation. For the simplest case where no water

> splitting occurs and the bipolar membrane acts like a simple electrodialysis memrbane these are:

Reviewer: ```suggestion
to fully solve the model. The exact degrees of freedom depend on the mode of operation. For the simplest case, where no water
splitting occurs and the bipolar membrane acts like a simple electrodialysis membrane, these are:
```

   "Membrane areal resistance", ":math:`r`", "membrane_surface_resistance", "['acidic', 'basic']", ":math:`\Omega m^2`", 2
   "Cell width", ":math:`b`", "cell_width", "None", ":math:`\text{m}`", 1
   "Cell length", ":math:`l`", "cell_length", "None", ":math:`\text{m}`", 1
   "Thickness of ion exchange membranes", ":math:`\delta`", "membrane_thickness", "['bpem']", ":math:`m`", 1
     "transport number of ions in the membrane phase", ":math:`t_j`", "ion_trans_number_membrane", "['bpem'], ['Na\ :sup:`+`', '\Cl\ :sup:`-`', 'H\ :sup:`+`', 'OH\ :sup:`-`']", "dimensionless", 4

**Note**
 :sup:`1` DOF number takes account of the indices of the corresponding parameter.

 :sup:`2` A user should provide either current or voltage as the electrical input, in correspondence to the "Constant_Current" or "Constant_Voltage" treatment mode

Solution component information
------------------------------
To fully construct solution properties, users need to provide basic component information of the feed solution to use this model. Below is a sample:

.. code-block::

   ion_dict = {
        "solute_list": ["Na_+", "Cl_-", "H_+", "OH_-"],
        "mw_data": {

>             "H2O": 18e-3,

Reviewer: ```suggestion
            "H2O": 18e-3,
```
Note, you no longer need to supply MW for water anymore.

            "Na_+": 23e-3,
            "Cl_-": 35.5e-3,
            "H_+": 1e-3,
            "OH_-": 17.0e-3,
        },
        "elec_mobility_data": {
            ("Liq", "Na_+"): 5.19e-8,
            ("Liq", "Cl_-"): 7.92e-8,
            ("Liq", "H_+"): 36.23e-8,
            ("Liq", "OH_-"): 20.64e-8,
        },
        "charge": {"Na_+": 1, "Cl_-": -1, "H_+": 1, "OH_-": -1},
        "diffusivity_data": {
            ("Liq", "Na_+"): 1.33e-9,
            ("Liq", "Cl_-"): 2.03e-9,
            ("Liq", "H_+"): 9.31e-9,
            ("Liq", "OH_-"): 5.27e-9,
        },
    }

> This model, by default, uses H\ :sub:`2`\ O  as the solvent of the feed solution.

Reviewer: I probably should wait until I get to the code, but what happens when the user omits H_+ and OH_- from the solute_list? Would the unit model raise the appropriate exception? If so, and I may have missed it, you should add a note here stating that H2 and OH must be provided.

Author: Yes, the user needs to supply  H_+ and OH_- . I have now added a note saying that it must be added.
